src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -595,39 +594,335 @@ public override bool IsTupleType
             }
         }
 
-        internal sealed override ObsoleteAttributeData ObsoleteAttributeData
+        public override bool IsImplicitlyDeclared
         {
             get
             {
-                return null;
+                return false;
+            }
+        }
+
+        public override ImmutableArray<TypeSymbol> TupleElementTypes
+        {
+            get
+            {
+                return _elementTypes;
+            }
+        }
+
+        public override ImmutableArray<string> TupleElementNames
+        {
+            get
+            {
+                return _elementNames;
             }
         }
 
         public override ImmutableArray<Symbol> GetMembers()
         {
-            return ImmutableArray<Symbol>.CastUp(_fields);
+            if (_lazyMembers.IsDefault)
+            {
+                ImmutableInterlocked.InterlockedInitialize(ref _lazyMembers, CreateMembers());
+            }
+
+            return _lazyMembers;
+        }
+
+        private ImmutableArray<Symbol> CreateMembers()
+        {
+            var namesOfVirtualFields = ArrayBuilder<string>.GetInstance(_elementTypes.Length);
+
+            if (_elementNames.IsDefault)
+            {
+                for (int i = 1; i <= _elementTypes.Length; i++)
+                {
+                    namesOfVirtualFields.Add(TupleMemberName(i));
+                }
+            }
+            else
+            {
+                namesOfVirtualFields.AddRange(_elementNames);
+            }
+
+            var members = ArrayBuilder<Symbol>.GetInstance(Math.Max(_elementTypes.Length, _underlyingType.OriginalDefinition.GetMembers().Length));
+
+            NamedTypeSymbol currentUnderlying = _underlyingType;
+            int currentNestingLevel = 0;
+
+            var fieldsForElements = ArrayBuilder<FieldSymbol>.GetInstance(currentUnderlying.Arity);
```

Comment:
Consider naming with prefix `current` since it will be updated with each loop iteration.